x/ecocredit/server/marketplace/buy.go

```diff
@@ -10,96 +10,21 @@ import (
-			}
-			batch, err := k.coreStore.BatchInfoTable().Get(ctx, sellOrder.BatchId)
-			if err != nil {
-
-				return nil, sdkerrors.ErrIO.Wrapf("error getting batch id %d: %s", sellOrder.BatchId, err.Error())
-			}
-			ct, err := utils.GetCreditTypeFromBatchDenom(ctx, k.coreStore, k.paramsKeeper, batch.BatchDenom)
-			if err != nil {
-				return nil, err
-			}
-			creditOrderQty, err := math.NewPositiveFixedDecFromString(order.Quantity, ct.Precision)
-			if err != nil {
-				return nil, err
-			}
-
-			// check that bid price and ask price denoms match
-			market, err := k.stateStore.MarketTable().Get(ctx, sellOrder.MarketId)
-			if err != nil {
-				return nil, fmt.Errorf("market id %d: %w", sellOrder.MarketId, err)
-			}
-			if order.BidPrice.Denom != market.BankDenom {
-				return nil, sdkerrors.ErrInvalidRequest.Wrapf("bid price denom does not match ask price denom: "+
-					" %s, expected %s", order.BidPrice.Denom, market.BankDenom)
-			}
-			// check that bid price >= sell price
-			sellOrderPricePerCredit, ok := sdk.NewIntFromString(sellOrder.AskPrice)
-			if !ok {
-				return nil, fmt.Errorf("could not convert %s to %T", sellOrder.AskPrice, sdk.Int{})
-			}
-			sellOrderPriceCoin := sdk.Coin{Denom: market.BankDenom, Amount: sellOrderPricePerCredit}
-			if sellOrderPricePerCredit.GT(order.BidPrice.Amount) {
-				return nil, ErrBidTooLow.Wrapf("sell order ask: %v, bid: %v", sellOrderPriceCoin, order.BidPrice)
-			}
-
-			// check address has the total cost (price per * order quantity)
-			bal := k.bankKeeper.GetBalance(sdkCtx, buyerAcc, order.BidPrice.Denom)
-			cost, err := getTotalCost(sellOrderPricePerCredit, order.Quantity)
-			if err != nil {
-				return nil, err
-			}
-			coinCost := sdk.Coin{Amount: cost, Denom: market.BankDenom}
-			if bal.IsLT(coinCost) {
-				return nil, sdkerrors.ErrInsufficientFunds.Wrapf("requested to purchase %s credits @ %s%s per "+
-					"credit (total %v) with a balance of %v", order.Quantity, sellOrder.AskPrice, market.BankDenom, coinCost, bal)
-			}
-
-			// fill the order, updating balances and the sell order in state
-			if err = k.fillOrder(ctx, sellOrder, buyerAcc, creditOrderQty, coinCost, false, !order.DisableAutoRetire, order.RetirementLocation, batch.BatchDenom); err != nil {
-				return nil, fmt.Errorf("error updating balances: %w", err)
-			}
-		case *marketplace.MsgBuy_Order_Selection_Filter:
-			return nil, sdkerrors.ErrInvalidRequest.Wrap("only direct buy orders are enabled at this time")
-			// verify expiration is in the future
-			//if order.Expiration.Before(sdkCtx.BlockTime()) {
-			//	return nil, sdkerrors.ErrInvalidRequest.Wrapf("expiration must be in the future: %s", order.Expiration)
-			//}
-		default:
-			return nil, sdkerrors.ErrInvalidRequest.Wrap("only direct buy orders are enabled at this time")
-		}
-	}
-	return &marketplace.MsgBuyResponse{}, nil
+	return nil, sdkerrors.ErrInvalidRequest.Wrap("only direct buy orders are enabled at this time")
```

Comment:
Should we do this in `ValidateBasic` rather than letting the message fail in the mempool?

Reply:
ah yeah good idea - fixed https://github.com/regen-network/regen-ledger/pull/967/commits/5f8029c324e18051550fbea5bb1360be95620ada